certbot/auth_handler.py

```diff
@@ -47,242 +42,144 @@ def __init__(self, auth, acme_client, account, pref_challs):
-            with error_handler.ExitHandler(self._cleanup_challenges, aauthzrs):
-                resp = self._solve_challenges(aauthzrs)
-                logger.info("Waiting for verification...")
+        authzrs = orderr.authorizations[:]
+        if not authzrs:
+            raise errors.AuthorizationError('No authorization to handle.')
+
+        # Retrieve challenges that need to be performed to validate authorizations.
+        achalls = self._choose_challenges(authzrs)
+        if not achalls:
+            return authzrs
+
+        # Starting now, challenges will be cleaned at the end no matter what.
+        with error_handler.ExitHandler(self._cleanup_challenges, achalls):
+            # To begin, let's ask the authenticator plugin to perform all challenges.
+            try:
+                resps = self.auth.perform(achalls)
+
+                # If debug is on, wait for user input before starting the verification process.
+                logger.info('Waiting for verification...')
+                config = zope.component.getUtility(interfaces.IConfig)
                 if config.debug_challenges:
+                    notify = zope.component.getUtility(interfaces.IDisplay).notification
                     notify('Challenges loaded. Press continue to submit to CA. '
                            'Pass "-v" for more info about challenges.', pause=True)
-
-                # Send all Responses - this modifies achalls
-                self._respond(aauthzrs, resp, best_effort)
-
-        # Just make sure all decisions are complete.
-        self.verify_authzr_complete(aauthzrs)
-
-        # Only return valid authorizations
-        ret_val = [aauthzr.authzr for aauthzr in aauthzrs
-                   if aauthzr.authzr.body.status == messages.STATUS_VALID]
-
-        if not ret_val:
-            raise errors.AuthorizationError(
-                "Challenges failed for all domains")
-
-        return ret_val
-
-    def _choose_challenges(self, aauthzrs):
+            except errors.AuthorizationError as error:
+                logger.critical('Failure in setting up challenges.')
+                logger.info('Attempting to clean up outstanding challenges...')
+                raise error
+            # All challenges should have been processed by the authenticator.
+            assert len(resps) == len(achalls), 'Some challenges have not been performed.'
+
+            # Inform the ACME CA server that challenges are available for validation.
+            for achall, resp in zip(achalls, resps):
+                self.acme.answer_challenge(achall.challb, resp)
+
+            # Wait for authorizations to be checked.
+            try:
+                self._poll_authorizations(authzrs, max_retries, best_effort)
+            except acme_errors.TimeoutError:
+                # Exceeding the max polling attempts, and some authentication are still not checked.
+                raise errors.AuthorizationError('All challenges could not be checked on time.')
```

Comment:
Yes I was not sure because I am not an english native speaker, but "checked" in french is "vérifié", and this word in french has a positive connotation: it usually means that the subject as made actions to be sure that the object fulfills the expectations. This is not about this here, we want to say that all authorizations are not in a definitive situation, so either valid or invalid, situation resolved by the ACME CA server.

I update accordingly to your suggestion.

Reply:
> it usually means that the subject as made actions to be sure that the object fulfills the expectations.

This is how the word is used in English as well. In this case, I was thinking of "checked" in the sense that the CA still hasn't checked/verified that the challenges were performed correctly, but that's not really describing the immediate problem.

I think the new wording is a lot better and hopefully you do too. Feel free to push back when I suggest wording changes (or any other changes). I'm certainly wrong sometimes.